addons/binding/org.openhab.binding.plclogo/src/main/java/org/openhab/binding/plclogo/discovery/PLCDiscoveryService.java

```diff
@@ -0,0 +1,189 @@
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.UID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link PLCDiscoveryService} is responsible for discovering devices on
+ * the current Network. It uses every Network Interface which is connected to a network.
+ *
+ * @author Alexander Falkenstern - Initial contribution
+ */
+public class PLCDiscoveryService extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(PLCDiscoveryService.class);
+
+    // Bridge config properties
+    private static final String LOGO_HOST = "address";
+
+    private static final int DISCOVERY_TIMEOUT = 30;
+    private static final Set<ThingTypeUID> DISCOVERABLE_THING_TYPES_UIDS = Collections.singleton(THING_TYPE_DEVICE);
+
+    private static final int CONNECTION_TIMEOUT = 500;
+    private TreeSet<String> addresses = new TreeSet<String>();
+
+    private ExecutorService executor = null;
+
+    private class Runner implements Runnable {
+        private final ReentrantLock lock = new ReentrantLock();
+        private final String host;
+
+        public Runner(final String address) {
+            Objects.requireNonNull(address, "IP may not be null");
+            this.host = address;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void run() {
+            // gets every ip which can be assigned on the given network
+            InetAddress address = null;
+            try {
+                address = InetAddress.getByName(host);
+                if (!address.isReachable(CONNECTION_TIMEOUT / 5)) {
+                    address = null;
+                }
+            } catch (IOException exception) {
+                logger.debug("LOGO! device not found at: {}.", host);
+                address = null;
+            }
+
+            final InetSocketAddress endpoint = new InetSocketAddress(host, 102);
+            if (!endpoint.isUnresolved() && (address != null)) {
+                Socket socket = new Socket();
+                try {
+                    socket.connect(endpoint, CONNECTION_TIMEOUT);
+                    logger.info("LOGO! device found at: {}.", host);
+
+                    String hostname = address.getHostName();
+                    if (!hostname.matches(UID.SEGMENT_PATTERN)) {
+                        // Replace invalid char's, since UID has no method for this.
+                        hostname = hostname.replaceAll("[^A-Za-z0-9_-]", "_");
+                    }
+
+                    final ThingUID thingUID = new ThingUID(THING_TYPE_DEVICE, hostname);
+                    DiscoveryResultBuilder builder = DiscoveryResultBuilder.create(thingUID);
+                    builder = builder.withProperty(LOGO_HOST, host);
```

Comment:
This assignment is not needed with a builder, you can even chain it for even nicer code.